src/core/annotations/qgsannotationitem.h

```diff
@@ -0,0 +1,155 @@
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSANNOTATIONITEM_H
+#define QGSANNOTATIONITEM_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgsrendercontext.h"
+#include "qgslinestring.h"
+#include "qgspolygon.h"
+
+class QgsFeedback;
+class QgsMarkerSymbol;
+class QgsLineSymbol;
+class QgsFillSymbol;
+
+/**
+ * \ingroup core
+ * Abstract base class for annotation items which are drawn with QgsAnnotationLayers.
+ *
+ * \since QGIS 3.16
+ */
+class CORE_EXPORT QgsAnnotationItem
+{
+
+#ifdef SIP_RUN
+    SIP_CONVERT_TO_SUBCLASS_CODE
+    if ( sipCpp->type() == "marker" )
+    {
+      sipType = sipType_QgsAnnotationMarkerItem;
+    }
+    else if ( sipCpp->type() == "linestring" )
+    {
+      sipType = sipType_QgsAnnotationLineStringItem;
+    }
+    else if ( sipCpp->type() == "polygon" )
+    {
+      sipType = sipType_QgsAnnotationPolygonItem;
+    }
+    else
+    {
+      sipType = 0;
+    }
+    SIP_END
+#endif
+
+  public:
+
+    /**
+     * Constructor for an annotation item, with the specified \a crs for storing
+     * its geographic location.
+     */
+    QgsAnnotationItem( const QgsCoordinateReferenceSystem &crs );
+
+#ifndef SIP_RUN
+    //! QgsAnnotationItem cannot be copied
+    QgsAnnotationItem( const QgsAnnotationItem &other ) = delete;
+    //! QgsAnnotationItem cannot be copied
+    QgsAnnotationItem &operator=( const QgsAnnotationItem &other ) = delete;
+#endif
+
+    virtual ~QgsAnnotationItem() = default;
+
+    /**
+     * Returns a clone of the item. Ownership is transferred to the caller.
+     */
+    virtual QgsAnnotationItem *clone() = 0 SIP_FACTORY;
+
+    /**
+     * Returns a unique (untranslated) string identifying the type of item.
+     */
+    virtual QString type() const = 0;
+
+    /**
+     * Returns the bounding box of the item's geographic location.
+     *
+     * The coordinate reference system for the bounding box is retrieved via crs().
+     */
+    virtual QgsRectangle boundingBox() const = 0;
+
+    /**
+     * Returns the CRS used for storing the location of the item.
+     *
+     * \see setCrs()
+     */
+    QgsCoordinateReferenceSystem crs() const { return mCrs; }
+
+    /**
+     * Sets the \a crs used for storing the location of the item.
+     *
+     * \see crs()
+     */
+    void setCrs( const QgsCoordinateReferenceSystem &crs );
```

Comment:
What's the reason for items having individual crs (rather than all being in layer crs)?

Reply:
Good question! I had my slyr hat on here, and annotations in ESRI world are stored using a referenced geometry + crs pair. So approaching it like this makes it easier to translate between the two. I don't see any problem in particular with this approach, expect a bit of extra code complexity. But I'm not set on it if you think it's a mistake...

Reply:
I was wondering what we gain (not sure if ESRI compatibility alone is convincing). I was a bit afraid of complexity: project crs vs layer crs vs item crs seems a lot of reprojection and possible room for missing one.

Reply:
Ah - but an annotation layer itself has no CRS. So it's just map CRS vs item CRS.

Reply:
Hmm, every layer has a `QgsMapLayer::crs()` by design.

Reply:
...But since a while we've supported returning invalid crs objects from this as a reflection of "no crs" 😁

Reply:
How are you going to work with this UI wise? Add it with the current project CRS? Will the CRS change if an item is moved?

Reply:
Ok, I'm convinced. I'll rework this!